### ISSUE TITLE
Difference between ex9 and ex18 matrices

### ISSUE THREAD

**Issue by vikramsg** [mfem/mfem#481]
I was looking at ex18 and was not sure how the derivative matrix for ex18 has been created. Essentially Aflux calculates the x and y derivatives without faces terms and adds them up. But the values seem incorrect. 

If we take periodic-square.mesh and we look at ex9 and print k matrix after removing face terms for velocity v(0) = 1; v(1) = 0, this is what we get

```
[row 77]
(77,-0.318764) (76,0.425019) (75,-0.106255)
[row 78]
(80,0.0664092) (79,-0.265637) (78,0.199228)
[row 79]
(80,-0.106255) (78,0.106255)
[row 80]
(80,-0.199228) (79,0.265637) (78,-0.0664092)
```

while printing Aflux gives

```
[row 77]
(161,-0.265637) (156,4.42374e-18) (155,0.265637) (77,0.318764)(76,0.170007) (75,-0.106255)
[row 78]
(159,0.199228) (156,0.106255) (153,-0.0664092) (80,0.0664092)(79,-0.106255) (78,-0.199228)
[row 79]
(160,0.318764) (159,-1.10594e-17) (157,0.170007) (154,-0.106255)(80,-0.265637) (78,0.265637)
[row 80]
(161,0.199228) (159,2.76484e-18) (158,0.106255) (155,-0.0664092)(80,0.199228) (79,0.106255) (78,-0.0664092)
```

Just comparing row 80, column 78, 79 and 80 should have the same values as k, but them seemed to have some values swapped from row 79.

**Comment by vikramsg** [comment]
After some further thought, I think you are using the weak form instead of the strong form DG. Is that correct?

**Comment by tzanio** [comment]
@jdahm, can you help Vikram with this issue?

**Comment by jdahm** [comment]
Yes, we're solving the weak form once integrated by parts. That is probably where the differences come from. Are there benefits that you see of solving the strong form instead?

**Comment by vikramsg** [comment]
@jdahm Thanks. I was trying to benchmark my code to yours and then saw the problem. 

Its actually mathematically equivalent, but I just think its conceptually simpler. But maybe you could include somewhere in the comments that it is using the weak form so that there's no confusion. Similarly maybe in ex9 it should be mentioned that it is using the strong form.

**Comment by v-dobrev** [comment]
@vikramsg, I'm not sure what you mean by "strong form" here - the method used in ex9 is obtained by integrating by parts twice, applying upwinding after the first integration by parts. I'm not aware of another formulation that will be "the weak form" - do you know of such a formulation?

Regarding ex18, it is a different problem, it is stated in conservative form (ex9 is not in conservative form), so in the derivation one has to integrate by parts only once and apply upwinding after that. I'm not aware of another way to derive (a stable) DG discretization. Can you clarify what you mean by "strong form" here?

In short, ex9 and ex18 are different equations, discretized using a similar approach but the resulting schemes are not directly comparable.

A version of ex9 that will be closer to ex18 is the equation:
```
du/dt + \div(\vec{v} u) = 0
```
discretized using the same approach as ex18.


Actually, the way I think of the derivation of the scheme used in ex9, is to first write it as
```
du/dt + \div(\vec{v} u) - \div(\vec{v}) u = 0
```
then apply the DG discretization to the first part
```
du/dt + \div(\vec{v} u)
```
including the upwinding. After that, combine the two volume terms (per element) into one
```
- (u, \vec{v} \cdot \grad(w)) - (u, \div{v} w) = - (u, \div(\vec{v} w))
```
And finally, integrate this term by parts, element by element, one more time.

**Comment by vikramsg** [comment]
[Nodal DiscontinuousGalerkin Methods1.pdf](https://github.com/mfem/mfem/files/2000151/Nodal.DiscontinuousGalerkin.Methods1.pdf)
[Nodal DiscontinuousGalerkin Methods2.pdf](https://github.com/mfem/mfem/files/2000152/Nodal.DiscontinuousGalerkin.Methods2.pdf)

I am attaching the two relevant pages from the book by Hesthaven, Warburton. Hopefully that is not violating any rules.

The second page attached shows the strong form, which is basically what you get when you integrate by parts twice as in ex9. Also, as stated in the same page, the two approaches are equivalent but the implementation will be different.

**Comment by v-dobrev** [comment]
@vikramsg, thanks for the clarification.

In ex9, using the weak form generally does not make sense because one needs `\div{\vec{v}}` in addition to `\vec{v}` itself. In the case of ex18, it seems that similar argument applies, unless the coefficients in the flux term (the coefficient a in the book) are constant. In your tests, do you have constant coefficients?

**Comment by vikramsg** [comment]
[Nodal DiscontinuousGalerkin Methods3.pdf](https://github.com/mfem/mfem/files/2004312/Nodal.DiscontinuousGalerkin.Methods3.pdf)

No. The strong form/weak form thing is valid for general equations of the conservative form. The above page shows it for a scalar equation in conservative form. Later in the book  they use it to do the Euler equations which do not have constant coefficients.

**Comment by v-dobrev** [comment]
If I understand equation (5.5) and the two expressions for the strong and weak forms correctly, it looks like the flux is being projected (in the same discrete space as the solution) before it is used. This is NOT what is done in ex9 - instead the flux is computed at the quadrature points. In the strong form, if you do not perform projection, you will need to know the derivatives of the flux to apply the chain rule to compute the residual at the quadrature points.

On the other hand, I think ex18 does perform the flux projection, even for the weak form. Is that correct @jdahm?

**Comment by jdahm** [comment]
When I responded originally that ex18 is using the weak form, I meant to give that as a reason that your compressible Euler example might not match the formulation used in my ex18, not as a different between ex9 and ex18.

@v-dobrev is correct that ex18 uses a flux projection. That is done in the call [here](https://github.com/mfem/mfem/blob/master/examples/ex18.hpp#L145), before multiplying it by the sparse matrix [here](https://github.com/mfem/mfem/blob/master/examples/ex18.hpp#L151), which computes `(f*, grad(w))` for all elements. Therefore, we use a flux projection, but it is on the second term in (5.4), not for the third (face) term.

**Comment by vikramsg** [comment]
I am not sure I understand this. So, in example 9, we have a vector u. Now to do the advection we do (M^(-1))K . x to get the RHS. This is excluding the face term. If we did a strong form evaluation of the Euler fluxes, we do the same thing. Only it becomes (M^(-1))K . f, where f is the flux. This is excluding the face terms. 

When you say, projection, do you mean f is computed as f = f(u(x)). Because anyway we would calculate it that way.

**Comment by v-dobrev** [comment]
The projection step is replacing (e.g. inside the integral) `f(u_h(x),x)` with `f_h(x)` where `f_h \in V_h` is a the projected flux, e.g. using nodal interpolation:
```
f_h(x) = \sum_i f(u_h(x_i), x_i) \ell_i(x)
```
where `\ell_i` are the nodal basis functions. This way, differentiating `f_h(x)` means differentiating `\ell_i(x)` whereas differentiating `f(u_h(x))` directly requires the derivative of `f` (and `u_h`).
